### DISCUSSION TITLE
How do I prevent ResourceLookups when unloading controls

### DISCUSSION THREAD

**Discussion by IAmTheCShark** [dotnet/wpf#5681] [Q&A]
We have some kind of "browser like" appication that is based on a TabControl that dynamically creates and removes tabpages, based on userinput.

TabPages sometimes contain complex views.

When removing tabpages we basically "cache" the content in a dictionary to quickly restore it, if it is needed again.
All this works fine, except that removing the TabPages sometimes takes a few seconds.
When hitting pause in the debugger ill allways land somewhere around these
![image](https://user-images.githubusercontent.com/20905599/141185793-942a2182-d06e-4119-b73c-0cc8ea161e2e.png)
So when a Controls VisualParent changes to null, it still causes a lot of invalidations of styles and resources.

I figured that moving things into Themes/Generic.xaml does improve the issue, but not completely prevent it.

So my question is, can I tell WPF to "relax" and not invalidate things? Or is there something I am doing wrong in general?

**Comment by IAmTheCShark** [comment]
Ok Im fairely certain now that this problem is worthy for an issue.

The StyleChanged events happen, when unloading controls, if you have "global styles" defined in App.xaml.
When the Parent / Visual Parent of these controls changes to null, they are no longer in the Visual nor logical tree of the App.xaml
and that causes the StyleProperty to reevaluate its effective value, since the source is now null (App.xaml resource).
So it defaults back, to the original style comming from a themes/generic.xaml.
this effectifely results in StyleChanged events when unloading controls.
The more controls and global styles, the worse the problem becomes.

It feels like those invalidations should not directly update the style, but wait for the control to enter the visual / logical tree. Then they would evauluate the same style resulting in no change.

**Reply by miloush** [reply]
I don't think elements not in the tree should have styles from a tree applied. An opt-in to supress unloading sounds like a better option to me.

**Reply by IAmTheCShark** [reply]
> I don't think elements not in the tree should have styles from a tree applied. An opt-in to supress unloading sounds like a better option to me.

You are right, that would be confusing. So optional suppressing sounds more reasonable.
